### PR TITLE
[WIP] [Do not merge] Fixes for some segmentation violations and FPEs 

### PR DESCRIPTION
**Pull request by andriish**
Hi @couet  and all,

this is a set of changes that should fix many of the recent crashes I've reported.

I'm marking this WIP as it should be up to you what should go to the master.

Best regards,
Andrii

### CODE REVIEW

File: graf2d/gpad/inc/TPad.h [hunk 1]
```diff
@@ -487,6 +487,7 @@ inline Int_t TPad::XtoAbsPixel(Double_t x) const
 //______________________________________________________________________________
 inline Int_t TPad::XtoPixel(Double_t x) const
 {
+   if (TMath::IsNaN(x)) return 0;
```
Comment (couet) [inline graf2d/gpad/inc/TPad.h:491]: Do you have a reproducer for that one ? such check in a such low level function might be a performance killer.....

Comment (andriish) [reply graf2d/gpad/inc/TPad.h:492]: Hi @cout ,

No, I don't have a standalone reproducer ( I see it in zevis) and the origin of nans seems to be elsewhere.


Also, I would suggest considering this branch as WIP and just a common place where I can put all the "kind of fixes" so you can easily see them.


Best regards,

Andrii

Comment (andriish) [reply graf2d/gpad/inc/TPad.h]: Found the problematic code. It was not in ROT, but just passed some garbage to ROOT functions. Remeved the fix.

**Comment by couet** [comment]
Note:

Such big PR is not good because they are hard to debug in case of problem. The recent recent example with the PR testing IsBatch in TCanvas::Close prove it. It was easy to track down because it was a single simple PR. This faulty code is present in this big PR. I suggest you remove this big PR. We cannot be confident with it.

**Comment by andriish** [comment]
Hi @couet, 

>This faulty code is present in this big PR. 

yes. That is why it is WIP. 

My idea is to chop big PR into smaller pieces and discuss/merge them.  But to keep bit PR so one can actually see the list of changes until there will be none.
Does it fit to the repository policy?  If not, if will remove it.

Best regards,
Andrii

**Comment by couet** [comment]
Well, we have other PRs tagged [WIP] ... so I guess that's fine.. I just do no like the idea of having clearly faulty code in a PR.

File: graf2d/gpad/src/TCanvas.cxx [hunk 2]
```diff
@@ -779,7 +779,7 @@ void TCanvas::Close(Option_t *option)
 {
    TPad    *padsave = (TPad*)gPad;
    TCanvas *cansave = 0;
-   if (!IsBatch()) if (padsave) cansave = (TCanvas*)gPad->GetCanvas();
+   if (padsave && gPad) cansave = (TCanvas*)gPad->GetCanvas();
```
Comment (couet) [inline graf2d/gpad/src/TCanvas.cxx:782]: This sounds stupid ...padsave is equal to gPad on line 780... did I missed soothing ?

Comment (pcanal) [reply graf2d/gpad/src/TCanvas.cxx:782]: Humm ... right .. I missed that ... this means that the original problem need to be debugged again to find the "real" problem.

**Comment by pcanal** [comment]
> The IsBatch() was logical

No it is not.  IsBatch "just" mean that there is attached screen but the code is still dealing with pad and canvas and current canvases (for example to store them in pdf file or ROOT file).

File: graf2d/gpad/src/TCanvas.cxx (hunk 2 above)
Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: It seems the fix is:
```
   if (padsave && gPad->TestBit(kNotDeleted)) cansave = (TCanvas*)gPad->GetCanvas();
```
I know this test is not always 100% sure but at least with this test, pad4 example dos not crash and stressGraphics is ok. And it is more logical as it is all about a double delete of the pad. @andriish can you check it solve your problem and make the appropriate PR ?

**Comment by andriish** [comment]
Closing this PR as the content was either merged or present in other PRs.
Thanks to all the participants!

**Pull request closed without merging**